describe a process for resolving image references in manifests (#31)
operator-framework/enhancements: enhancements/image-references.md

is slightly different and some of the semantics about how the output should be 
formatted are dissmilar, the core image reference rewriting is identical and the
process of building a `FROM scratch` image layer is also identical.

Today, some prior art exists in the OSBS workflow for building operator bundles.
As we improve the Operator SDK tooling to create a straghtforward process for
creating bundle images, we must make sure a seamless migration is possible.

### Risks and Mitigations

It will be critical that the design be vetted by all of the concerned parties, from
operator manifest authors to CI system authors and productized pipeline authors to
ensure that the UX is appropriate in all cases. Furthermore, the largest risk in the
implementation here is not prioritizing a clean migration pathway for all current
users who create bundle images, which would lead to further fragmentation of the
ecosystem, which is directly opposed to the goal of this enhancement.

## Design Details

The definition of a minimally-viable operator bundle image will be changed to
ensure that all iamge references in the contained manifests have been resolved 

> to a digest and had the pull specifications rewritten to refer to those digests.

Reviewer: I think this constraint should be opt-in for upstream users and mandatory for admittance to catalogs for k8s distros that require it.

Author: I disagree. A bundle with unresolved image references is not reproducible and will be hard to support in any context it's used. If the bundle creation tooling makes it trivial to resolve a spec to a digest, what use would someone have for an unresolved spec in a bundle?

Reviewer 2: @stevekuznetsov to clarify: are you saying that you want all bundles to include `image-references` and `image-replacements`?

Reviewer 3: I'm not sure I see the argument here - how are `image-references` and `image-replacements` useful for users that don't have the split build systems / pipelines that openshift has?

An upstream user would build the operator/operand images, record their digests in their bundle manifests, and publish it.

Author: If an upstream user still wants to create reproducible artifacts they still want to SHA-pin. I agree at that point though image digest resolution could be a simple "look up the current digest for this floating tag" without an indirection, so the `image-replacements` would be inferred.

The only acceptable process for creating an operator bundle will be to run the

> `operator-sdk bundle create` CLI, providing the manifests, metadata and image

Reviewer 4: Aren't bundle images created via `opm` today? How does that fit into the process?

Reviewer 5: Bundle images can be created by an alpha version of the `opm` command or the operator-sdk command. 

The SDK command is where we should focus this. `opm` is for when you add the bundle to the catalog, which is the next step for the pipelines to do once the bundle is created.

sources as input to the creation process.

### Proposed UX

Operator manifest authors write a manifest that refers to images using some
opaque string [ex](https://github.com/openshift/machine-config-operator/blob/master/install/0000_80_machine-config-operator_04_deployment.yaml#L22),

> and provide an `image-references` file alongside their manifests that declares

Reviewer: The current [on-disk bundle spec](https://github.com/openshift/enhancements/blob/master/enhancements/olm/operator-bundle.md#example-of-an-operator-bundle-that-uses-operator-registry-format) defines a `metadata` directory where arbitrary metadata can be added. 

Maybe we could extend the current spec with some new fields in the `metadata/annotations.yaml` -- or define a new file -- to handle this mapping.

I would also like to make the point that a "kubectlable" resource NOT intended for cluster application could be considered misleading.

Author: I would like to keep the implementation as identically similar to that used in `oc adm release new` as possible. Any explicitly different semantics will contribute to maintenance load on the shared library that will execute this logic. I don't have a strong opinion on this being a new file in the `metadata/` directory but for interoperability if it were `image-references` and co-located with manifests (as it is in OpenShift payloads) that would be simplest.

Reviewer 2: A new file makes sense to me, since `annotations.yaml` contains annotations that must appear on a bundle image as `LABEL`s.

which strings inside of their manifest are referring to pull specifications of
images and names each occurence [ex](https://github.com/openshift/machine-config-operator/blob/98d9ba6841eb4811ed6f4d4de7016ea83c131c54/install/image-references#L7-L10).

> The `image-references` file holds data in the format of an OpenShift `ImageStream`:

Reviewer 3: It would be nice (but not a blocker) if we could define this mapping without the ImageStream type, so that we could advertise this as a way upstream users can support their own CI workflows.

> The `image-references` file holds data in the format of an OpenShift `ImageStream`:

Author: As I note here: https://github.com/operator-framework/enhancements/pull/31#discussion_r443714730

Any deviations in this workflow from that which is used in `oc adm release new` comes at significant cost and should not be considered lightly.

```yaml
kind: ImageStream
apiVersion: image.openshift.io/v1
spec:
  tags:
  - name: my-image-name # this is the semantic identifier for this image
    from:
      kind: DockerImage
      name: registry.svc.ci.openshift.org/openshift:image # this is an opaque string which exists in manifests
```

This mapping, therefore, defines what needs to be replced in manifests when
they are bundled and identifies each replacement with a name. When
`operator-sdk bundle create` runs, it will require as input a second mapping
from those names to literal image pull specifications and will run the replacement.
In this manner, the configuration provided by the manifest author remains static
regardless of the eventual replacement that a build system will execute.

The second mapping required at build-time will also take the form of an `ImageStream`
and will be provided either via reading an `ImageStream` from an OpenShift cluster or
reading a serialized `ImageStream` from disk. A build system that does not store the
images used as sources in an `ImageStream` may therefore provide a static file instead.
The mapping to provide the image sources looks like:

```yaml
apiVersion: image.openshift.io/v1

> kind: ImageStreamImageam

Author: Remove the trailing `m`

status:
  tags:
  - tag: my-image-name # this matches the semantic identifier provided by the manifest author
    items:
    - dockerImageReference: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:298d0b496f67a0ec97a37c6e930d76db9ae69ee6838570b0cd0c688f07f53780 # this is the fully-resolved pull specification that will be used in replacements
      image: sha256:298d0b496f67a0ec97a37c6e930d76db9ae69ee6838570b0cd0c688f07f53780
```

> `operator-sdk bundle create` will use this chain of mapping to perform replacements

Reviewer 5: It may be worth calling out that this image is not needed to be multi-arch. As well that the image will have a `media-type: application/vnd.docker.distribution.manifest.v2+json` and be conforming to the version 2 schema 2 spec?

in the manifests before creating a bundle image layer. The layer creation will also
be shared logic with `oc adm release new` in order to allow both processes to build

> image layers without requiring the use of a containe runtime, other build system, any

Reviewer 5: nit: s/containe/container

---
title: image-references-in-operator-bundles
authors:
  - "@stevekuznetsov"
reviewers:
  - "@jwforres"
  - "@shawn-hurley"
  - "@gallettilance"

>   - "@lui"

Author: Yes :)

approvers:
  - "@jwforres"
  - "@shawn-hurley"
creation-date: 2020-05-17
last-updated: 2020-05-17
status: implementable
see-also:
  - "/enhancements/olm/operator-bundle.md"
---

# Image References in Operator Bundle Manifests

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

Manifests that make up an Operator Bundle for installation via the Operator
Lifecycle Manager refer to one or more container images by their pull

> specifications: these container images define the operator and operands that

Reviewer 6: Isn't the operand only referred to by resources the operator creates?

Author: Not sure i understand this question.

Reviewer 3: Any images required by the operator (for operands, etc) need to be surfaced at the packaging layer for mirroring to disconnected environments.

the manifests deploy and manage. As with OpenShift release payloads, operator
bundles must refer to images by digest in order to produce reproducible

> installations. A shared process to build operator bundles that replaces image

Author: This sounds totally reasonable, allowing someone to do both is a good idea.

references with fully-resolved pull specifications that pin images by digest
must be built; this process must allow for a number of separate build systems
to direct how these replacements occur in order to support a full-featured
build and test strategy.

## Motivation

### Goals

> - there is one, canonical, method for building an operator bundle image layer

Reviewer 2: Is this desirable, considering that we want the bundle format to be an open standard?

Author: Why not?

Reviewer 2: Counter question: is there a good reason to say that `operator-sdk` is the only tool that should build bundle images (and/or image layers, which are proposed as an addition to the spec by this EP)? The spec is well-documented and there's a KEP, so any tool should be able to build a bundle image.

I'm effectively saying that this goal could be removed without any other changes to the EP.

Author: Perhaps this could be narrowed to specify that there is one such process for any bundles created within Red Hat build systems or productized pipelines. The standard and output format continues to be open and if someone else were to want to implement a separate process for building such an image layer, that should continue to be possible. From a consensus/maintenance perspective, however, Red Hat should have no interest in seeing separate processes exist to create bundles in our various internal build systems.

Reviewer 2: I agree with that, but this is an upstream enhancement proposal and not specific to Red Hat so this goal should be specified elsewhere.

Author: What if we change `s/one/a/` in this line? I explicitly do want this to be a canonical method.

Author: (and I think if we do not produce something that is trivial to operate with and does not become ~ubiquitous we have failed)

> - there is one, canonical, method for building an operator bundle image layer

Author: Yes

  from a directory of manifests

> - building an operator bundle image does not require the use of a container

Author: `tar` :)

> - building an operator bundle image does not require the use of a container

Reviewer 7: How do we turn a tarball into something that can be pushed to a registry?

> - building an operator bundle image does not require the use of a container

Reviewer 7: I don't seem to have access to reopen this, but my previous question has not been answered.

  runtime, elevated privileges or any capacities that are not present for 
  containerized workloads on OpenShift
- it is as simple for a developer to build a bundle referring to test versions
  of operand images as it is for a CI or productized build pipleine to create
  bundle images for publication

> - operator manifest authors dictate the set of image references in manfiests

Author: I disagree. This proposal, for the `operator-sdk` ecosystem, expects authors that opt into image digest pinning to do so explicitly. A tool may be built to adapt the current `$RELATED_IMAGE_foo` and related mechanisms as used by OSBS into an `image-references` file but I do not necessarily think that it is generically applicable to the ecosystem.

> - operator manifest authors dictate the set of image references in manfiests

Reviewer 8: nit: "manfiests" -> "manifests"

  that must be resolved and pinned
- operator bundle images may be inspected to determine the pull specifications
  that were used in the creation of the bundle

### Non-Goals

> - operator manifest authors must not be required to define the registry from

Author: Both counts, explicitly yes.

  which any individual build system will resolve image references

> - upstream operator manifests must not be required to know how common names or

Author: I hope the examples clarify this. It's an identifier that relies neither on the image reistry, namespace name or tag.

>   referecnes change when built in a downstream pipeline

Reviewer 2: ```suggestion
  references change when built in a downstream pipeline
```

- no prescriptive statement is made about the specific format or contents of the
  bundle image layer; any will be transparently supported as an output

## Proposal

### User Stories

#### Story 1

As an author of a manifest, I would like to check in manifests to my upstream
repository that are self-consistent, valid and make no assumptions about the
build system that will eventually create a bundle image with them.

#### Story 2

As an author of an operand, I would like to create a bundle locally in order
to test my operator end-to-end on a cluster of my choosing without having to
edit the core configuration for the operator.

#### Story 3

As an author of a build system, I would like to operate with tooling that allows
me to clearly define the source of truth for image digests in order to keep the
build-system-specific configuration to an absolute minimum.

#### Story 4

> As an engineer involved in publishing an optional operator, I would like to

Author: The user must declare what they expect to be replaced, the build system determines how to do so.

configure semantically equivalent image pull specifications once, in order to not
need to configure each build system independently.

### Implementation Details/Notes/Constraints

The core problems that must be solved in the implementation of this proposal have
already been handled in the workflow used in `oc adm release new`. When implementing

> improvements to the `operator-sdk bundle create` process we will simply need to

Reviewer 2: What part of `operator-sdk bundle create` does this proposal plan on improving? If it is the metadata generation part, this should be updated to
```suggestion
improvements to the `operator-sdk generate bundle` process we will simply need to
```

Author: Sorry, this is changing too fast to keep up with ...

create a shared library for the two tools to use. While the shape of the output
is slightly different and some of the semantics about how the output should be 
formatted are dissmilar, the core image reference rewriting is identical and the
process of building a `FROM scratch` image layer is also identical.

Today, some prior art exists in the OSBS workflow for building operator bundles.
As we improve the Operator SDK tooling to create a straghtforward process for
creating bundle images, we must make sure a seamless migration is possible.

### Risks and Mitigations

It will be critical that the design be vetted by all of the concerned parties, from
operator manifest authors to CI system authors and productized pipeline authors to
ensure that the UX is appropriate in all cases. Furthermore, the largest risk in the
implementation here is not prioritizing a clean migration pathway for all current
users who create bundle images, which would lead to further fragmentation of the
ecosystem, which is directly opposed to the goal of this enhancement.

## Design Details

The definition of a minimally-viable operator bundle image will be changed to

> ensure that all iamge references in the contained manifests have been resolved 

Reviewer 2: ```suggestion
ensure that all image references in the contained manifests have been resolved 
```

> to a digest and had the pull specifications rewritten to refer to those digests.

Reviewer 7: An image reference that needs pinning can appear in multiple locations in the CSV file:

1. RelatedImage: `spec.relatedImages[].image`
2. Annotation: any of:
   1. `metadata.annotations.containerImage`
   2. `metadata.annotations[]` for any given Object where the annotation value contains one
      or more image pull specification.
3. Container: `spec.install.spec.deployments[].spec.template.spec.containers[].image`
4. InitContainer: `spec.install.spec.deployments[].spec.template.spec.initContainers[].image`
5. RelatedImageEnv: `spec.install.spec.deployments[].spec.template.spec.containers[].env[].value`
   and `spec.install.spec.deployments[].spec.template.spec.initContainers[].env[].value` where the
   `name` of the corresponding env Object is prefixed by `RELATED_IMAGE_`.

This is the currently supported locations for IIB, OSBS, and CVP.

In order to replace an image reference with its pinned form, all of those locations must be searched. 
I don't think there's a way to do the replacements without being aware of those locations. A simple `sed-like` approach will definitely not cover all use cases. `2.ii` is particularly difficult.

I bring up this point because from an implementation POV the process in this proposal **must** be aware of each one of those locations. Otherwise, it cannot replace image references reliably.

It's desirable to have the list of valid image reference locations in a centralized location, currently we have at least 3.

This proposal **must** include a mechanism which can be used to identify which image references are in the CSV.

Reviewer 7: If I understand the proposal correctly, you could get away with not knowing about those locations because we're forcing authors to be explicit about image references. IMO, we're moving complexity from the tooling to the authors which seems like a step backwards. For instance, an author now needs to provide the same placeholder pull spec in `RELATED_IMAGE_FOO` environment variable, again in `spec.relatedImages`, and one more time in image-references.

Author: The explicit ask from authors that wish to opt into image pinning pays large dividends by making a simple, transparent and interoperable system. The fact that you just pasted a long list of different ways for someone to explicitly or implicitly opt into it today, and the fact that the resolution process would be duplicated across build pipelines speaks to that. This is an active and explicit decision.

Author: > For instance, an author now needs to provide the same placeholder pull spec in RELATED_IMAGE_FOO environment variable, again in spec.relatedImages, and one more time in image-references.

Wheras today ... they need to provide what is effectively a placeholder that happens to also be an image pull spec. I don't see the functional difference here.

Reviewer 7: Consider this example that the build systems can handle today:

```yaml
# csv file
metadata:
  annotations:
    containerImage: registry.example.com/image:tag
```

Once it's processed via the build system, it's transformed into something like:

```yaml
# csv file
metadata:
  annotations:
    containerImage: registry.example.com/image@sha256:123...
spec:
  relatedImages:
    - name: some-name
      value: registry.example.com/image@sha256:123...
```

The authors only have to define the pull spec once.

With this proposal, the author would have to write this:
```yaml
# csv file
metadata:
  annotations:
    containerImage: registry.example.com/image:tag
spec:
  relatedImages:
    - containerImage: registry.example.com/image:tag

# image-references
imageReferences:
- name: common-name
  placeholder: registry.example.com/image:tag
```

Authors have to duplicate the placeholder pull spec at least three times for every unique pull spec across 2 different files. Additionally, authors must now have some sort of static analysis to make sure there are no typos in those pull specs. Otherwise, they may not be replaced as desired, which may be easy to miss if there are 10 different pull spec in the CSV file.

As I mentioned earlier, this is moving the complexity from the build system to the authors.

I do like the idea of removing the current "magic" of finding pull specs in all sorts of places in the CSV file. What if instead of placeholder pull specs, we used an actual templating language? The author could write the following:

```yaml
# csv file
metadata:
  annotations:
    containerImage: {{ COMMON_NAME }}
spec:
  relatedImages:
    - name: common-name
      value: {{ COMMON_NAME }}

# image-references
imageReferences:
- name: COMMON_NAME
  # default value if image-replacements is not provided.
  default: registry.example.com/image:tag
```

We could even go a step further and provide an option to automatically populate `spec.relatedImages` from image-references.

Author: You are inventing the requirement for authors to replcate it three times. The proposal as written does not require it. We ask authors to identify their _common names_ for images and the opaque identifier that this common name is represented as. This is all they need to do:

```yaml
# image-references
imageReferences:
- name: common-name
  placeholder: registry.example.com/image:tag
```

You are pushing too much in the wrong direction. Please understand that this explicit mapping between an identifier and the placeholder is critical for interoperability between build systems. This allows the CSV to remain entirely static. If we tie the identifier to the registry/org/name/tag that the CSV happens to use, it is implicit and confusing when you add into the picture:

 - registry renames that occur differently between build systems
 - organization changes that apply to separate registries
 - _name_ changes (for instance, the Bugzilla component mapping that OSBS does)
 - tag changes, as when an upstream `latest` could become a downstream `:release-3.4` or whatever

Please also understand that the "standard" that OSBS/IIB/CVP is using today is not a proposal for the `operator-sdk` ecosystem, as this one is. It does not provide a platform for the community at large. It is not ubiquitous nor is it applicable generally.

By explicitly expecting the user to provide a manifest that identifies the unique identifiers and their placeholders in the CSV we create an explicit, clear and straightforward interface for _all_ build systems. As you can (I hope) clearly see from the long list of implicit, semi-explicit, etc means by which OSBS currently supports an author making effectively the same decision, the state today is not simple. It is not interoperable. How might a user, today, configure that their CSV refers to an operand which may have one name upstream but another once it's processed by OSBS and renamed for a Bugzilla component? By using _common names_ which apply to the underlying identifier, a user may propose that their CSV relies on that common name _regardless of how the build system wants to rename it at build-time_.

If a build system would like to apply implicit heuristics for finding image pull specifications and provide their users with a pre-populated `image-references`, that build system may continue to do so. It will then incur _all_ of the complexity and cost that you are finding yourself managing today. The approach as used by the OpenShift release payload scales, is clear, explicit and transparent. A static analysis check to ensure that the user resolved all references is trivial to implement and CSV authors incur the cost once in the lifetime of their CSV.

Reviewer 7: > You are inventing the requirement for authors to replcate it three times.

Given the image-references file you provided in your previous comment, could you also post the corresponding CSV file? Both must be taken into account because both are maintained by the authors. I'm really curious as to how `registry.example.com/image:tag` does not appear three times across both files in the case where disconnected support is needed.

Maybe it wasn't clear but the point I'm trying to highlight is the fact that manually maintaining placeholder pull specs across two different files is error prone. Tooling must be available to help authors with this. This is true regardless if you're building a bundle image locally or in a build system.

Author: There is no difference between placing an opaque string in three places or a template language placeholder. They hold identical error cases. CI that builds your bundle, installs your operator and verifies it will catch those trivially.

The only acceptable process for creating an operator bundle will be to run the
`operator-sdk bundle create` CLI, providing the manifests, metadata and image
sources as input to the creation process.

### Proposed UX

Operator manifest authors write a manifest that refers to images using some

> opaque string [ex](https://github.com/openshift/machine-config-operator/blob/master/install/0000_80_machine-config-operator_04_deployment.yaml#L22),

Reviewer 2: Can we be clear about the wording here and say `image tag`? Saying `opaque string`, while general, implies that you can set this to anything which could break a Deployment that is used both in and outside of a bundle.

Author: This is explicitly not an image tag. Why should we care to support the Deployment outside of a Bundle?

Reviewer 2: I have an operator project that I would like to bundle. I modify my Deployment's operator image tag to be an opaque string. Now I can bundle my operator successfully, but running `kubectl apply -f deployment.yaml` breaks unless I keep 2 versions of my Deployment or modify it after generating a CSV. We should not be breaking existing operator projects for image references.

Instead, I should be able to leave my image tag as-is, which is effectively:
```yaml
imageReferences:
- name: registry.svc.ci.openshift.org/openshift:image
  placeholder: registry.svc.ci.openshift.org/openshift:image
```

So we can simplify the image references spec to:
```yaml
imageReferences:
- registry.svc.ci.openshift.org/openshift:image
```
because wouldn't the author want all references to the same tag be updated in the same way? This string is just as easy to reference in `image-replacements`.

Author: Why do we care to support `kubectl apply`? The entire point of the operator is to deploy all of the manifests together. Why should a developer want to apply some small subset of their manifests instead of running a (trivially short) bundle creation and deploying that way, even for their development work?

Author: I explicitly do not want to see us guessing at the common names here. Allowing an explicit mapping between a common name and the opaque string is required to interoperate between separate build systems which may call their images by different names, but need a common name as currency to uniquely identify the container image regardless.

Author: Furthermore -- the opqaue string may happen to be a valid pull spec, so if a packager were to use an upstream set of manifests verbatim, there would be no backwards-incompatible change from this proposal and no breaking for existing operator projects.

Reviewer 2: `kubectl apply` was just an example. Given a new operator project, I should be able to run
```sh
$ kustomize build config/default | kubectl apply -f -
```
which includes all operator manifests required to run my operator.

The explicit mapping is fine, I'm mainly arguing against saying this is an opaque string because it can be a valid pull spec which is not opaque. Given your last point, `placeholder` could be optional in case you're using a valid pull spec. Then the first example above would look like:
```yaml
imageReferences:
- name: registry.svc.ci.openshift.org/openshift:image
```

Maybe we're just arguing over syntax here.

Author: > Given a new operator project, I should be able to run

Why?

Reviewer 2: This is what [`make deploy`](https://github.com/operator-framework/operator-sdk-samples/blob/master/go/kubebuilder/memcached-operator/Makefile#L46) does in the [new project layout](https://master.sdk.operatorframework.io/docs/golang/quickstart/#build-and-run-the-operator). `operator-sdk` needs to be a drop-in replacement for `kubebuilder` while also supporting bundles, so the ability to do this can/should not change.

Author: Makes sense. Keeping backwards compatibility here should be simple, and I agree we should strive to do so.

> opaque string [ex](https://github.com/openshift/machine-config-operator/blob/master/install/0000_80_machine-config-operator_04_deployment.yaml#L22),

Reviewer 6: If we have a link like this, lets please pin it to a commit for posterity

Reviewer 6: Also, there is no reason to have a link here, just put the literal example in here

> opaque string [ex](https://github.com/openshift/machine-config-operator/blob/master/install/0000_80_machine-config-operator_04_deployment.yaml#L22),

Reviewer 7: This sounds like templating with extra steps. It seems error prone to duplicate the placeholder image reference in multiple locations.

Author: No less error prone than writing a pull spec in those places that all point to the same image, though.

and provide an `image-references` file alongside their manifests that declares
which strings inside of their manifest are referring to pull specifications of
images and names each occurence [ex](https://github.com/openshift/machine-config-operator/blob/98d9ba6841eb4811ed6f4d4de7016ea83c131c54/install/image-references#L7-L10).

The `metadata/image-references` file holds data in the following format, mapping

> common names of container imags to their string placeholders in the manifests:

Reviewer 7: `s/imags/images`

```yaml
imageReferences:
- name: common-name

>   placeholder: registry.svc.ci.openshift.org/openshift:image

Reviewer 3: Quick bikeshed on `placeholder`, what about aligning somewhat with `sed` terminology, like:

 - `substitute`
 - `pattern`
 
and do we think we will only ever allow full-text matching?

Author: `substitute` also sounds good, I don't feel strongly on this.

Reviewer 8: We don't want to use [an `ImageStream`][1] like [OpenShift release images do][2]?

[1]: https://github.com/openshift/machine-config-operator/blob/915ca7a11fead41d75ffa0ed7e310f7284645a13/install/image-references
[2]: https://github.com/openshift/cluster-version-operator/blob/62ff8e3e51cd4bfe6cd3b8cb6400a28ada1b183d/docs/dev/operators.md#how-do-i-ensure-the-right-images-get-used-by-my-manifests

Author: That was proposed at the start, but an OpenShift-agnostic approach was suggested by the maintainers. I think a translation library from an ImageStream to these more targeted formats is simple to write, so there is no harm in using them. As factored today, a list of `struct`s allows us room in the future to expand the API surface in a simple but backwards compatible manner, too.

Reviewer 8: Talk about that choice in an *Alternatives* section, so folks curious about the divergence don't have to dig it out of PR discussion?

```

> This mapping, therefore, defines what needs to be replced in manifests when

Reviewer 2: ```suggestion
This mapping, therefore, defines what needs to be replaced in manifests when
```

they are bundled and identifies each replacement with a name. When

> `operator-sdk bundle create` runs, it will require as input a second mapping

Reviewer 2: >as input

This would need to be a file/stored in a config somewhere, since typing all this out in an interactive prompt or in flag args would be painful.

Author: Yes, agreed.

from those names to literal image pull specifications and will run the replacement.
In this manner, the configuration provided by the manifest author remains static
regardless of the eventual replacement that a build system will execute.

> The second mapping required at build-time, known as `image-replacements`, will take

Reviewer 3: is the name of the file important? or is it only important that there be a file with a top level `imagePullSpecs` field?

if `image-replacements` is required to be yaml, we should call it `image-replacements.yaml`, or if not allow both `yaml` and `json` suffices.

Author: I simply followed the pattern from the OpenShift release payload process which uses a suffix-less file. I don't feel strongly here and am happy to use anything that we have consensus over.

Reviewer 6: +1 for `.yaml` to allow IDEs to point out invalid content

a similar form, mapping the common names of images to their explicitly resolved pull
specifications. This file will have the following format:

```yaml
imagePullSpecs:
- name: common-name
  pullSpec: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:298d0b496f67a0ec97a37c6e930d76db9ae69ee6838570b0cd0c688f07f53780
```

`operator-sdk generate bundle` will use this chain of mapping to perform replacements
in the manifests before creating a bundle image layer. The layer creation will also
be shared logic with `oc adm release new` in order to allow both processes to build

> image layers without requiring the use of a containe runtime, other build system, any

Reviewer 3: ```suggestion
image layers without requiring the use of a container runtime, other build system, any
```

> elevated permissions, privleges, capacities or SELinux roles. As the output image

Reviewer 2: ```suggestion
elevated permissions, privileges, capacities or SELinux roles. As the output image
```

layer in both cases is `FROM scratch` and simply contains manifest data, this build
process is simple and producing the layer by creating the underlying tar bundle does
not come with risks.

The bundle creation process will furthermore create as output an `images` metadata file
that will expose the build-time inputs used to create the bundle, so that downstream
consumers can access this data as necessary. This file will also allow the build system
to add any further metadata necessary. The format for the `images` file is a strict

> super-set of the `image-replacements` file:

Reviewer 2: Why can't `image-replacements` be used directly here? What "metadata" do you envision being required that isn't stored elsewhere.

Author: @smarterclayton @wking @masuzlik could expand on how this is used for OpenShift release payloads.

Reviewer 4: maszulik is @soltysh

Reviewer 8: Do we have this for OpenShift release images?  We have [a separate `release-metadata` file in those][1], maybe you're just making space for additional metadata like that in a single file?

The most important metadata we store in `release-metadata` is the release image's version string (e.g. `4.1.0`).  Having that inside the signed release image means that consumers can say "would you like to update to 4.1.0?" without having to worry about whether the by-digest pullspec they'll use is actually for some different release.  But if you always bundle into container images you could theoretically store that metadata as [labels][2] on the image too.

[1]: https://github.com/openshift/cluster-version-operator/blob/62ff8e3e51cd4bfe6cd3b8cb6400a28ada1b183d/docs/user/reconciliation.md#release-image-content
[2]: https://github.com/opencontainers/image-spec/blame/v1.0.1/config.md#L159-L162

Author: Ah, I was mistaken in my understanding, then. I'll update to use a `release-metadata` file.

references with fully-resolved pull specifications that pin images by digest
must be built; this process must allow for a number of separate build systems
to direct how these replacements occur in order to support a full-featured
build and test strategy.

## Motivation

### Goals

- there is one, canonical, method for building an operator bundle images
  from a directory of manifests
- it is possible to perform image resolution and pinning separate from bundle
  creation, but not the opposite
- building an operator bundle image does not require the use of a container
  runtime, elevated privileges or any capacities that are not present for 
  containerized workloads on OpenShift
- it is as simple for a developer to build a bundle referring to test versions
  of operand images as it is for a CI or productized build pipleine to create
  bundle images for publication
- operator manifest authors dictate the set of image references in manfiests
  that must be resolved and pinned
- operator bundle images may be inspected to determine the pull specifications
  that were used in the creation of the bundle

### Non-Goals

> - operator manifest authors must not be required to define the registry from

Reviewer 6: That seems to be a goal, not a `non-goal`

Reviewer 8: How about leading with an explicit "The following statements limit the scope of this enhancement:"?  I'm also fine with "The following are not goals of this enhancement:" and your suggested sense reversal on the list entries.

Reviewer 8: Or lead each entry like "This enhancement will not require operator manifest authors..." or some such.

Reviewer 6: Thats still unclear. `must not ` can par definition not describe a non-goal. A non-goal describes work that should not be done, not work that should be done to avoid a problem. To me its unclear if this here should be done or not.

Author: No, it's explicitly a non-goal. The author defines a name. The build system connects that to a location.

Reviewer 6: If its a requirement that the author doesn't define the registry, then this is definitely a goal, not a non-goal.

  which any individual build system will resolve image references

> - upstream operator manifests must not be required to know how common names or

Reviewer 6: Same here, either this is a goal in disguise or poorly worded. If this is actually not a goal, it should be something like
```
* Requiring upstream operator manifests to know how common names or references change when built in a downstream pipeline
```

Author: They should not be involved in the "how" of resolution in _any_ part of the build system.

Reviewer 6: okay so it is a goal? `Must` and `Must not` can never describe the absence of a goal (aka `non-goal`). If something should explicitly be avoided, thats a goal, not a non-goal.

  references change when built in a downstream pipeline

> - no prescriptive statement is made about the specific format or contents of the

Reviewer 6: This also sounds like a goal to me

Author: Why? This has no bearing whatsoever on the format of the files in the bundle.

ensure that the UX is appropriate in all cases. Furthermore, the largest risk in the
implementation here is not prioritizing a clean migration pathway for all current
users who create bundle images, which would lead to further fragmentation of the
ecosystem, which is directly opposed to the goal of this enhancement.

## Design Details

The definition of a minimally-viable operator bundle image will be changed to
ensure that all image references in the contained manifests have been resolved 
to a digest and had the pull specifications rewritten to refer to those digests.

The only acceptable process for creating an operator bundle will be to run the
`operator-sdk generate bundle` CLI, providing the manifests, metadata and image
sources as input to the creation process.

### Proposed UX

Operator manifest authors write a manifest that refers to images using some
opaque string [ex](https://github.com/openshift/machine-config-operator/blob/master/install/0000_80_machine-config-operator_04_deployment.yaml#L22),
and provide an `image-references` file alongside their manifests that declares
which strings inside of their manifest are referring to pull specifications of

> images and names each occurence [ex](https://github.com/openshift/machine-config-operator/blob/98d9ba6841eb4811ed6f4d4de7016ea83c131c54/install/image-references#L7-L10).

Reviewer 6: First there is a link to an example, the below that is a literal example, why?

Reviewer 6: Also the link example is plain wrong, according to the discusison below where its said that an `ImageStream` is not supposed to be used

Author: Yeah we should remove this link

In this manner, the configuration provided by the manifest author remains static
regardless of the eventual replacement that a build system will execute.

The second mapping required at build-time, known as `image-replacements.yaml`, will take
a similar form, mapping the common names of images to their explicitly resolved pull
specifications. This file will have the following format:

```yaml
imagePullSpecs:
- name: common-name
  pullSpec: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:298d0b496f67a0ec97a37c6e930d76db9ae69ee6838570b0cd0c688f07f53780
```

`operator-sdk generate bundle` will use this chain of mapping to perform replacements
in the manifests before creating a bundle image layer. The layer creation will also
be shared logic with `oc adm release new` in order to allow both processes to build
image layers without requiring the use of a container runtime, other build system, any
elevated permissions, privileges, capacities or SELinux roles. As the output image
layer in both cases is `FROM scratch` and simply contains manifest data, this build
process is simple and producing the layer by creating the underlying tar bundle does
not come with risks.

> The bundle creation process will commit the `image-replacements.yaml` file into the output

Reviewer 5: which directory in the bundle will these files be in?

  bundle image layers; any will be transparently supported as an output

## Proposal

### User Stories

#### Story 1

As an author of a manifest, I would like to check in manifests to my upstream
repository that are self-consistent, valid and make no assumptions about the
build system that will eventually create a bundle image with them.

#### Story 2

As an author of an operand, I would like to create a bundle locally in order
to test my operator end-to-end on a cluster of my choosing without having to
edit the core configuration for the operator.

#### Story 3

As an author of a build system, I would like to operate with tooling that allows
me to clearly define the source of truth for image digests in order to keep the
build-system-specific configuration to an absolute minimum.

#### Story 4

As an engineer involved in publishing an optional operator, I would like to
configure semantically equivalent image pull specifications once, in order to not
need to configure each build system independently.

### Implementation Details/Notes/Constraints

The core problems that must be solved in the implementation of this proposal have

> already been handled in the workflow used in `oc adm release new`. When implementing

Author: I think that is likely the best source of information

manifests that declares which strings inside of their manifest are referring
to pull specifications of images and names each occurence.

The `metadata/image-references.yaml` file holds data in the following format, mapping
common names of container images to their string placeholders in the manifests:

```yaml
imageReferences:
- name: common-name
  substitute: registry.svc.ci.openshift.org/openshift:image
```

This mapping, therefore, defines what needs to be replaced in manifests when
they are bundled and identifies each replacement with a name. When
`operator-sdk generate bundle` runs, it will require as input a second mapping
from those names to literal image pull specifications and will run the replacement.
In this manner, the configuration provided by the manifest author remains static
regardless of the eventual replacement that a build system will execute.

The second mapping required at build-time, known as `image-replacements.yaml`, will take
a similar form, mapping the common names of images to their explicitly resolved pull
specifications. This file will have the following format:

```yaml
imagePullSpecs:
- name: common-name

>   pullSpec: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:298d0b496f67a0ec97a37c6e930d76db9ae69ee6838570b0cd0c688f07f53780

Reviewer 7: In most current cases, the maintainer provides a pull spec to a tag which the build system resolves to a digest. Having the build system do the digest pinning greatly improves the user experience. 

I'm wondering how the build system could make the jump from `registry.svc.ci.openshift.org/openshift:image` to `quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:298d0b496f67a0ec97a37c6e930d76db9ae69ee6838570b0cd0c688f07f53780`. It seems like we need an intermediate pullspec, e.g. `quay.io/openshift-release-dev/ocp-v4.0-art-dev:latest`.

One possibility is for maintainers, the ones using the build system, to provide a non-digest pullspec in `image-replacements.yaml`. Then, the build system could simply resolve the unresolved pullspecs in it at build time.

Does that make sense?

Author: I don't think there is any jump required here - based on how the `ImageStream` in `registry.svc.ci.openshift.org` is configured, the normal mechanism for determining image digests will resolve the layers either at the underlying registry - quay, in this case - or the OpenShift registry as configured. 

I *do* expect maintainers of the build system to change the non-digest pullspec in `image-replacements.yaml` to point to the "right thing" - the build system should know, given a semantic name, where it should live. For instance, CI will know that `my-operator` semantically means a tag in an internal registry. OSBS will know that `my-operator` is owned by a specific `dist-git` repo, etc (the config for this is already something you ask users to provide IIRC) and will know whether the build should use e.g. the staging or prod reigstry. The build system would set values as necessary and the build would resolve them to a digest by simply following the normal rules and expecting the value being resolved to be authoritative.

Reviewer 7: `ImageStream` is not listed anywhere in this doc. Did you mean `ImageReference`, `ImagePullSpec`, or something else?

I think we're on the same page regarding how `image-replacements.yaml` will be used by a build system. The maintainer creates it to point to some agreed upon location, e.g. a floating tag in the production registry, then the build system is responsible for altering the pull specs in `image-replacements.yaml`, "on the fly" as part of the build process, to use a digest for example.

Author: `ImageStream` being the underlying technology that you end up using when you hit that particular registry, using the registry API, IIRC. Meaning, it's up to the admin who sets up the registry and not something the user cares about here.

```

`operator-sdk generate bundle` will use this chain of mapping to perform replacements
in the manifests before creating a bundle image layer. The layer creation will also
be shared logic with `oc adm release new` in order to allow both processes to build
image layers without requiring the use of a container runtime, other build system, any
elevated permissions, privileges, capacities or SELinux roles. As the output image
layer in both cases is `FROM scratch` and simply contains manifest data, this build
process is simple and producing the layer by creating the underlying tar bundle does
not come with risks.

The bundle creation process will commit the `image-replacements.yaml` file into the output
bundle, and will furthermore create as output a `release-metadata.yaml` metadata file
that will expose any further build-time inputs used to create the bundle, so that downstream
consumers can access this data as necessary. The format for the `release-metadata.yaml` file
is a loose key-value store:

```yaml
metadata:
  some-key: some-value
```

The `image-replacments.yaml` content will also be injected into the CSV using the extant

> `relatedImages` stanza for backwards compatibility, but this data is not expected to be

Reviewer 7: Yes, populating `relatedImages` automatically is a great direction.

Could you clarify:
1. What if `relatedImages` is already defined in the CSV in source control? Is it overwritten, merged, or is an error raised?
1. If the answer to the above is yes, will image replacements in a pre-existing `relatedImages` block be supported?
1. There could be cases where populating `relatedImages` is not desirable, but image replacements are. Would there be an option to turn off populating the `relatedImages` block?

Perhaps this is too low level for this design.

Author: 1. The current implementation of `relatedImages` ([doc](https://github.com/operator-framework/operator-lifecycle-manager/blob/4edf102c1054b9e7e8de772f9d0316425cca6513/doc/contributors/design-proposals/related-images.md#associating-related-images)) provides a mechanism for OLM to know which images must be mirrored when users install into a disconnected environment, that is all. That mechanism is slightly related to the current proposal but ultimately has nothing to do with it. I don't see why a user would _want_ to provide any existing `relatedImages` if they can be handled automatically by the tooling, so I would propose disallowing a user to set `relatedImages` explicitly.
2. Image replacements in an existing `relatedImages` are not a thing - there is no `operator-sdk` tooling that does this. I don't follow the question.
3. I don't understand how this case can be possible - if the operator deployment needs an image to exist for it to work, it must exist in `relatedImages` for this deployment to function in a disconnected environment, and not adding that image to `relatedImages` would not allow that. Could you please explain the use-case you are envisioning?

Author: @lcarva if you are talking in 2 about the current functionality for `containerbuildsystem/operator-manifest` this conversation (and that about forgetting a field) will be solved in the glue logic that adaps current CSV input from users in `dist-git` to use this mechanism.

Reviewer 7: 1. Raising an error if `relatedImages` is already defined in the CSV in source control sounds like a reasonable option here. Could you please update the doc to explicitly state this?
1. Based on your answer to 1, question 2 is no applicable. Let's disregard it. 
1. The use case I had in mind is if disconnected deployments are not yet supported for a given operator. I'm not entirely sure what it means to actually support disconnected deployments. Is it just populating `relatedImages`? Or is there additional work, e.g. testing, that needs to be done?

Author: ACK - will update for 1 & 2.

For 3 - just checked and I *think* it's in CVP somewhere that you are able to mark your operator supported in a disconnected manner, and that triggers a check that `spec.relatedImages` is set. So setting it regardless should be safe.

- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

Manifests that make up an Operator Bundle for installation via the Operator
Lifecycle Manager refer to one or more container images by their pull
specifications: these container images define the operator and operands that
the manifests deploy and manage. As with OpenShift release payloads, operator
bundles must refer to images by digest in order to produce reproducible
installations. A shared process to build operator bundles that replaces image
references with fully-resolved pull specifications that pin images by digest
must be built; this process must allow for a number of separate build systems
to direct how these replacements occur in order to support a full-featured
build and test strategy.

## Motivation

### Goals

- there is one, canonical, method for building an operator bundle images
  from a directory of manifests
- it is possible to perform image resolution and pinning separate from bundle
  creation, but not the opposite
- building an operator bundle image does not require the use of a container

>   runtime, elevated privileges or any capacities that are not present for 
>   containerized workloads on OpenShift

Reviewer 2: ```suggestion
  runtime, elevated privileges or any capacities that are available in a
  high-security context
```

As an author of a manifest, I would like to check in manifests to my upstream
repository that are self-consistent, valid and make no assumptions about the
build system that will eventually create a bundle image with them.

#### Story 2

As an author of an operand, I would like to create a bundle locally in order
to test my operator end-to-end on a cluster of my choosing without having to
edit the core configuration for the operator.

#### Story 3

As an author of a build system, I would like to operate with tooling that allows
me to clearly define the source of truth for image digests in order to keep the
build-system-specific configuration to an absolute minimum.

#### Story 4

As an engineer involved in publishing an optional operator, I would like to
configure semantically equivalent image pull specifications once, in order to not
need to configure each build system independently.

### Implementation Details/Notes/Constraints

The core problems that must be solved in the implementation of this proposal have
already been handled in the workflow used in `oc adm release new`. When implementing
improvements to the `operator-sdk generate bundle` process we will simply need to

> create a shared library for the two tools to use. While the shape of the output

Reviewer 9: Since this enhancement proposes breaking out a shared library which is currently part of `oc` I'm raising this to the attention of @soltysh in case you haven't already reviewed it.

Reviewer 10: I'm fine with the shared code, library-go comes to my mind immediately since we share some of the release code with CVO there, already, so this seems like a nice expansion of that functionality.

is slightly different and some of the semantics about how the output should be 
formatted are dissmilar, the core image reference rewriting is identical and the
process of building a `FROM scratch` image layer is also identical.

Today, some prior art exists in the OSBS, IIB and CVP workflows for building
operator bundles. As we improve the Operator SDK tooling to create a straghtforward
process for creating bundle images, we must make sure a seamless migration is possible.

### Risks and Mitigations

It will be critical that the design be vetted by all of the concerned parties, from

> operator manifest authors to CI system authors and productized pipeline authors to

Reviewer 9: AIUI there is a key difference between what is proposed here and how IIB currently works, which is that this method finds all references in the manifests directory and replaces them all. The existing IIB process only replaces in specific locations. I definitely prefer the approach in this enhancement as it gives more flexibility to operator authors and behaves the same way as openshift's image replacement used for its core payload building (giving us consistency in developer experience). However as far as some pipelines are concerned, it does mean it could potentially be replacing strings that were not being replaced before. Have we reached out to any operator authors yet to gauge whether this is going to be a concern?

common names of container images to their string placeholders in the manifests:

```yaml
imageReferences:
- name: common-name
  substitute: registry.svc.ci.openshift.org/openshift:image
```

This mapping, therefore, defines what needs to be replaced in manifests when
they are bundled and identifies each replacement with a name. When
`operator-sdk generate bundle` runs, it will require as input a second mapping
from those names to literal image pull specifications and will run the replacement.
In this manner, the configuration provided by the manifest author remains static
regardless of the eventual replacement that a build system will execute.

The second mapping required at build-time, known as `image-replacements.yaml`, will take
a similar form, mapping the common names of images to their explicitly resolved pull
specifications. This file will have the following format:

```yaml
imagePullSpecs:
- name: common-name
  pullSpec: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:298d0b496f67a0ec97a37c6e930d76db9ae69ee6838570b0cd0c688f07f53780
```

`operator-sdk generate bundle` will use this chain of mapping to perform replacements

> in the manifests before creating a bundle image layer. The layer creation will also

Reviewer 9: For devs running the command themselves, will there be an option to `--skip-image-replacements` ? I'm thinking about while an operator author is actively developing, maybe they do just want to have some tag references so they can push updates to the images without having to rebuild the bundle every time.

be shared logic with `oc adm release new` in order to allow both processes to build
image layers without requiring the use of a container runtime, other build system, any
elevated permissions, privileges, capacities or SELinux roles. As the output image
layer in both cases is `FROM scratch` and simply contains manifest data, this build
process is simple and producing the layer by creating the underlying tar bundle does
not come with risks.

The bundle creation process will commit the `image-replacements.yaml` file into the output
bundle, and will furthermore create as output a `release-metadata.yaml` metadata file
that will expose any further build-time inputs used to create the bundle, so that downstream
consumers can access this data as necessary. The format for the `release-metadata.yaml` file
is a loose key-value store:

```yaml
metadata:
  some-key: some-value

> ```

Reviewer 9: Understand this is intended as open-ended metadata, but providing examples of what a build system might include here would provide more context.
